File: 0047-permutations-ii/0047-permutations-ii.java
class Solution {
    public List<List<Integer>> permuteUnique(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        List<Integer> current = new ArrayList<>();
        Arrays.sort(nums);
        backtrack(result, current, nums, new boolean[nums.length]);
        return result;
    }
    
    public void backtrack(List<List<Integer>> result, List<Integer> current, int[] nums, boolean[] used){
        if(current.size()==nums.length){
            result.add(new ArrayList<>(current));
            return;
        }
        for(int i=0;i<nums.length;i++){
            if(used[i] || (i > 0 && nums[i] == nums[i-1] && !used[i-1])){
                continue;
            }
            used[i] = true;
            current.add(nums[i]);
            backtrack(result, current, nums, used);
            current.remove(current.size()-1);
            used[i] = false;
        }
        
        
    }
}

File: 0047-permutations-ii/0047-permutations-ii.py
class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        result = []
        nums.sort()
        def backtrack(used, current):
            if len(current) == len(nums):
                result.append(current.copy())
                return
            for i in range(0, len(nums)):
                if used[i] or (i > 0 and nums[i] == nums[i-1] and not used[i-1]):
                    continue
                current.append(nums[i])
                used[i] = True
                backtrack(used, current)
                used[i] = False
                current.pop()
        backtrack([False]*len(nums), [])
        return result
                
        